{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Probe Match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/junyuchen/Lab/Probe-Match'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prokka - Contigs annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prokka --outdir mydir --prefix mygenome contigs.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prokka --outdir Root11 --prefix Root11 Root11.fna"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```shell\n",
    "for fasta in `ls | grep .fna`\n",
    "do\n",
    "    prokka --outdir /home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/${fasta:0:-4} --prefix ${fasta:0:-4} $fasta\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Parising GFF file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### bcbio-gff\n",
    "conda install -c bioconda bcbio-gff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from BCBio import GFF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/junyuchen/Lab/Probe-Match'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "from BCBio import GFF\n",
    "\n",
    "in_file = \"your_file.gff\"\n",
    "\n",
    "limit_info = dict(\n",
    "        gff_id = [\"chr1\"],\n",
    "        gff_source = [\"Coding_transcript\"])\n",
    "\n",
    "in_handle = open(in_file)\n",
    "for rec in GFF.parse(in_handle, limit_info=limit_info):\n",
    "    print rec.features[0]\n",
    "in_handle.close()\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "product=23S ribosomal RNA\n",
    "product=16S ribosomal RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID = contig_1, length 1914680, with 0 features\n",
      "[]\n",
      "ID = contig_2, length 1264344, with 0 features\n",
      "[]\n",
      "ID = contig_3, length 926414, with 3 features\n",
      "[SeqFeature(FeatureLocation(ExactPosition(67862), ExactPosition(67971), strand=-1), type='rRNA', id='MOFEJJPE_03177'), SeqFeature(FeatureLocation(ExactPosition(68060), ExactPosition(70869), strand=-1), type='rRNA', id='MOFEJJPE_03178'), SeqFeature(FeatureLocation(ExactPosition(71675), ExactPosition(73160), strand=-1), type='rRNA', id='MOFEJJPE_03181')]\n",
      "ID = contig_4, length 409661, with 0 features\n",
      "[]\n",
      "ID = contig_5, length 209741, with 0 features\n",
      "[]\n",
      "ID = contig_6, length 59929, with 0 features\n",
      "[]\n",
      "ID = contig_7, length 14163, with 0 features\n",
      "[]\n",
      "ID = contig_8, length 5484, with 0 features\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "#in_file = \"/home/junyuchen/Lab/Probe-Match/Root-WGS/Root11/Root11.gff\"\n",
    "in_file = \"/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root123D2/Root123D2.gff\"\n",
    "\n",
    "limit_info = dict(gff_type = [\"rRNA\"])\n",
    "        \n",
    "\n",
    "in_handle = open(in_file)\n",
    "for record in GFF.parse(in_handle, limit_info=limit_info):\n",
    "    #print(type(rec.features))\n",
    "    #print(rec.features[0].qualifiers[\"product\"])\n",
    "    #print(rec.id)\n",
    "    #print(rec.features[0])\n",
    "    print(\"ID = %s, length %i, with %i features\"\n",
    "          % ( record.id, len(record.seq), len(record.features)))\n",
    "    print (record.features)\n",
    "    \n",
    "in_handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_file = \"/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root123D2/Root123D2.gff\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "type: rRNA\n",
      "location: [208745:210295](-)\n",
      "id: NLKIOJMO_05069\n",
      "qualifiers:\n",
      "    Key: ID, Value: ['NLKIOJMO_05069']\n",
      "    Key: locus_tag, Value: ['NLKIOJMO_05069']\n",
      "    Key: product, Value: ['16S ribosomal RNA']\n",
      "    Key: score, Value: ['0']\n",
      "    Key: source, Value: ['barrnap:0.9']\n",
      "\n",
      "contig_9\n",
      "type: rRNA\n",
      "location: [205678:208597](-)\n",
      "id: NLKIOJMO_05068\n",
      "qualifiers:\n",
      "    Key: ID, Value: ['NLKIOJMO_05068']\n",
      "    Key: locus_tag, Value: ['NLKIOJMO_05068']\n",
      "    Key: product, Value: ['23S ribosomal RNA']\n",
      "    Key: score, Value: ['0']\n",
      "    Key: source, Value: ['barrnap:0.9']\n",
      "\n",
      "contig_9\n",
      "-1\n",
      "205678\n",
      "208597\n",
      "<class 'Bio.SeqFeature.FeatureLocation'>\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "in_file = \"/home/junyuchen/Lab/Probe-Match/Root-WGS/Root11/Root11.gff\"\n",
    "\n",
    "limit_info = dict(gff_type = [\"rRNA\"])\n",
    "        \n",
    "\n",
    "in_handle = open(in_file)\n",
    "for record in GFF.parse(in_handle, limit_info=limit_info):\n",
    "    #print(type(rec.features))\n",
    "    #print(rec.features[0].qualifiers[\"product\"])\n",
    "    #print(rec.id)\n",
    "    #print(rec.features[0])\n",
    "    #print(\"ID = %s, length %i, with %i features\"\n",
    "    #      % ( record.id, len(record.seq), len(record.features)))\n",
    "    #print(record.features)\n",
    "    #print (len(record.features))\n",
    "    if len(record.features) != 0:\n",
    "        for i in range(len(record.features)):\n",
    "            #print(record.id)\n",
    "            #print(record.features[0])\n",
    "            #print(len(record.seq))\n",
    "            #print(record.features[i-1].qualifiers[\"product\"])\n",
    "            #print(type(record.features[i-1].qualifiers[\"product\"]))\n",
    "            if record.features[i-1].qualifiers[\"product\"] == ['23S ribosomal RNA']:\n",
    "                print(record.features[i-1])\n",
    "                print(record.id)\n",
    "                print(record.features[i-1].location.strand)\n",
    "                print(record.features[i-1].location.start)\n",
    "                print(record.features[i-1].location.end)\n",
    "                print(type(record.features[i-1].location))\n",
    "                #feature_seq = record.seq[record.features[i-1].location.start:record.features[i-1].location.end].reverse_complement()\n",
    "                #print(feature_seq)\n",
    "                seq_23S = SeqRecord(record.features[i-1].location.extract(record.seq),\\\n",
    "                                    id=record.features[i-1].id,\\\n",
    "                                    description=str(record.id +'-'+ record.features[i-1].qualifiers[\"product\"][0]))\n",
    "                \n",
    "                #print(seq_23S)\n",
    "                SeqIO.write(seq_23S, \"my_example.faa\", \"fasta\")\n",
    "            elif record.features[i-1].qualifiers[\"product\"] == ['16S ribosomal RNA']:\n",
    "                print(record.features[i-1])\n",
    "                print(record.id)\n",
    "                seq_16S = record.features[i-1].location.extract(record.seq)\n",
    "                #print(seq_16S)\n",
    "                \n",
    "                \n",
    "            #print\n",
    "        #print(record.features)\n",
    "in_handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/junyuchen/Lab/Probe-Match'"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "feature_seq = example_parent[example_feature.location.start:example_feature.location.end].reverse_complement()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "feature_seq = example_feature.extract(example_parent)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "contig_9\tbarrnap:0.9\trRNA\t205679\t208597\t0\t-\t.\tID=NLKIOJMO_05068;locus_tag=NLKIOJMO_05068;product=23S ribosomal RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root920/Root920\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root920\n",
      "Root920\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root265/Root265\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root265\n",
      "Root265\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root630/Root630\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root630\n",
      "Root630\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root70/Root70\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root70\n",
      "Root70\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root473/Root473\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root473\n",
      "Root473\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1298/Root1298\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1298\n",
      "Root1298\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root495/Root495\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root495\n",
      "Root495\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root170/Root170\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root170\n",
      "Root170\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root166/Root166\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root166\n",
      "Root166\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root708/Root708\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root708\n",
      "Root708\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root83/Root83\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root83\n",
      "Root83\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1279/Root1279\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1279\n",
      "Root1279\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root149/Root149\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root149\n",
      "Root149\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1240/Root1240\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1240\n",
      "Root1240\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1280/Root1280\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1280\n",
      "Root1280\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root901/Root901\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root901\n",
      "Root901\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root123D2/Root123D2\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root123D2\n",
      "Root123D2\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root482/Root482\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root482\n",
      "Root482\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1294/Root1294\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1294\n",
      "Root1294\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root133/Root133\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root133\n",
      "Root133\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root9/Root9\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root9\n",
      "Root9\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1212/Root1212\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1212\n",
      "Root1212\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root627/Root627\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root627\n",
      "Root627\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root4/Root4\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root4\n",
      "Root4\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root491/Root491\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root491\n",
      "Root491\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root401/Root401\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root401\n",
      "Root401\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root614/Root614\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root614\n",
      "Root614\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root268/Root268\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root268\n",
      "Root268\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1497/Root1497\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1497\n",
      "Root1497\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root483D1/Root483D1\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root483D1\n",
      "Root483D1\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1310/Root1310\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1310\n",
      "Root1310\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root236/Root236\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root236\n",
      "Root236\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1252/Root1252\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1252\n",
      "Root1252\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1203/Root1203\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1203\n",
      "Root1203\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1464/Root1464\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1464\n",
      "Root1464\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root52/Root52\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root52\n",
      "Root52\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root562/Root562\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root562\n",
      "Root562\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root11/Root11\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root11\n",
      "Root11\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root11/.ipynb_checkpoints/Root11-checkpoint\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root11/.ipynb_checkpoints\n",
      "Root11-checkpoint\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root569/Root569\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root569\n",
      "Root569\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root241/Root241\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root241\n",
      "Root241\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1485/Root1485\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1485\n",
      "Root1485\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1485/.ipynb_checkpoints/Root1485-checkpoint\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1485/.ipynb_checkpoints\n",
      "Root1485-checkpoint\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root214/Root214\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root214\n",
      "Root214\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root561/Root561\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root561\n",
      "Root561\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root68/Root68\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root68\n",
      "Root68\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1471/Root1471\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1471\n",
      "Root1471\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root137/Root137\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root137\n",
      "Root137\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root318D1/Root318D1\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root318D1\n",
      "Root318D1\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root217/Root217\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root217\n",
      "Root217\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root136/Root136\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root136\n",
      "Root136\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root672/Root672\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root672\n",
      "Root672\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root186/Root186\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root186\n",
      "Root186\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root186/.ipynb_checkpoints/Root186-checkpoint\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root186/.ipynb_checkpoints\n",
      "Root186-checkpoint\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1423/Root1423\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1423\n",
      "Root1423\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root444D2/Root444D2\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root444D2\n",
      "Root444D2\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root135/Root135\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root135\n",
      "Root135\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1204/Root1204\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1204\n",
      "Root1204\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1312/Root1312\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root1312\n",
      "Root1312\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root122/Root122\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root122\n",
      "Root122\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root71/Root71\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root71\n",
      "Root71\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root483D2/Root483D2\n",
      "/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/Root483D2\n",
      "Root483D2\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "dir = \"/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/\"\n",
    "\n",
    "for root,dirs,files in os.walk(dir):\n",
    "    #print(dirs)\n",
    "    for tempFile in files:\n",
    "        if \"gff\" in tempFile:\n",
    "            path = os.path.join(root,tempFile)\n",
    "            print(path[:-4])\n",
    "            #print(dirs)\n",
    "            print(root)\n",
    "            print(tempFile[:-4])\n",
    "            \n",
    "            \n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "limit_info = dict(gff_type = [\"rRNA\"])\n",
    "        \n",
    "def FindrRNA(path, file):\n",
    "    \n",
    "    in_handle = open(path)\n",
    "\n",
    "    for record in GFF.parse(in_handle, limit_info=limit_info):\n",
    "\n",
    "        if len(record.features) != 0:\n",
    "            for i in range(len(record.features)):\n",
    "\n",
    "                if record.features[i-1].qualifiers[\"product\"] == ['23S ribosomal RNA']:\n",
    "                    #print(record.features[i-1])\n",
    "                    #print(record.id)\n",
    "                    #print(record.features[i-1].location.strand)\n",
    "                    #print(record.features[i-1].location.start)\n",
    "                    #print(record.features[i-1].location.end)\n",
    "                    #print(type(record.features[i-1].location))\n",
    "                    #feature_seq = record.seq[record.features[i-1].location.start:record.features[i-1].location.end].reverse_complement()\n",
    "                    #print(feature_seq)\n",
    "                    seq_23S = SeqRecord(record.features[i-1].location.extract(record.seq),\\\n",
    "                                    id=record.features[i-1].id,\\\n",
    "                                    description=str(record.id +'-'+ record.features[i-1].qualifiers[\"product\"][0]))\n",
    "                \n",
    "                    #print(seq_23S)\n",
    "                    #SeqIO.write(seq_23S, path[:-4]+\"_23S_rRNA.fasta\", \"fasta\")\n",
    "                    SeqIO.write(seq_23S, \"/home/junyuchen/Lab/Probe-Match/Root-rRNA/\"+tempFile[:-4]+\"_23S_rRNA.fasta\", \"fasta\")\n",
    "                elif record.features[i-1].qualifiers[\"product\"] == ['16S ribosomal RNA']:\n",
    "                    #print(record.features[i-1])\n",
    "                    #print(record.id)\n",
    "                    seq_16S = SeqRecord(record.features[i-1].location.extract(record.seq),\\\n",
    "                                    id=record.features[i-1].id,\\\n",
    "                                    description=str(record.id +'-'+ record.features[i-1].qualifiers[\"product\"][0]))\n",
    "                \n",
    "                    #SeqIO.write(seq_16S, path[:-4]+\"_16S_rRNA.fasta\", \"fasta\")\n",
    "                    SeqIO.write(seq_16S, \"/home/junyuchen/Lab/Probe-Match/Root-rRNA/\"+tempFile[:-4]+\"_16S_rRNA.fasta\", \"fasta\")\n",
    "                    #print(seq_16S)\n",
    "\n",
    "\n",
    "                #print\n",
    "            #print(record.features)\n",
    "    in_handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from BCBio import GFF\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "\n",
    "dir = \"/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/\"\n",
    "\n",
    "for root,dirs,files in os.walk(dir):\n",
    "    #print(dirs)\n",
    "    for tempFile in files:\n",
    "        if \"gff\" in tempFile:\n",
    "            path = os.path.join(root,tempFile)\n",
    "            FindrRNA(path, tempFile)\n",
    "            #print(path[:-4])\n",
    "            #print(dirs)\n",
    "            #print(root)\n",
    "            #print(tempFile[:-4])\n",
    "            \n",
    "            \n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "Created on Sat Nov  2 09:52:16 2019\n",
    "\n",
    "@author: Junyu\n",
    "\"\"\"\n",
    "\n",
    "import os\n",
    "from BCBio import GFF\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "\n",
    "dir = \"/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/\"\n",
    "\n",
    "limit_info = dict(gff_type = [\"rRNA\"])\n",
    "        \n",
    "def FindrRNA(path, tempFile):\n",
    "    \n",
    "    in_handle = open(path)\n",
    "\n",
    "    for record in GFF.parse(in_handle, limit_info=limit_info):\n",
    "\n",
    "        if len(record.features) != 0:\n",
    "            for i in range(len(record.features)):\n",
    "\n",
    "                if '23S ribosomal RNA' in record.features[i].qualifiers[\"product\"][0]:\n",
    "                    #print(record.features[i])\n",
    "                    #print(record.id)\n",
    "                    #print(record.features[i].location.strand)\n",
    "                    #print(record.features[i].location)\n",
    "                    #print(record.features[i].location.start)\n",
    "                    #print(record.features[i].location.end)\n",
    "                    #print(type(record.features[i].location))\n",
    "                    #feature_seq = record.seq[record.features[i].location.start:record.features[i].location.end].reverse_complement()\n",
    "                    #print(feature_seq)\n",
    "                    '''\n",
    "                    seq_23S = SeqRecord(record.features[i].location.extract(record.seq),\\\n",
    "                                    id=record.features[i].id,\\\n",
    "                                    description=str(record.id +'-'+ record.features[i].qualifiers[\"product\"][0]))\n",
    "                    '''\n",
    "                    loc23 = int(record.features[i].location.end) - int(record.features[i].location.start)\n",
    "                    #print(loc)\n",
    "                    seq_23S = SeqRecord(record.features[i].location.extract(record.seq),\\\n",
    "                                    id=tempFile[:-4],\\\n",
    "                                    description=str(record.id +'-'+record.features[i].id+'-'+ record.features[i].qualifiers[\"product\"][0]+'-'+str(loc23)+ \"bp\"))\n",
    "                    #print(seq_23S)\n",
    "                    #print(record.features[i].location.start)\n",
    "                    #SeqIO.write(seq_23S, path[:-4]+\"_23S_rRNA.fasta\", \"fasta\")\n",
    "                    SeqIO.write(seq_23S, \"/home/junyuchen/Lab/Probe-Match/Root-rRNA2/\"+tempFile[:-4]+\"_23S_rRNA.fasta\", \"fasta\")\n",
    "                elif '16S ribosomal RNA' in record.features[i].qualifiers[\"product\"][0]:\n",
    "                    #print(record.features[i])\n",
    "                    #print(record.id)\n",
    "                    loc16 = int(record.features[i].location.end) - int(record.features[i].location.start)\n",
    "                    seq_16S = SeqRecord(record.features[i].location.extract(record.seq),\\\n",
    "                                    id=tempFile[:-4],\\\n",
    "                                    description=str(record.id +'-'+record.features[i].id+'-'+ record.features[i].qualifiers[\"product\"][0]+'-'+str(loc16)+\"bp\"))\n",
    "                \n",
    "                    #SeqIO.write(seq_16S, path[:-4]+\"_16S_rRNA.fasta\", \"fasta\")\n",
    "                    SeqIO.write(seq_16S, \"/home/junyuchen/Lab/Probe-Match/Root-rRNA2/\"+tempFile[:-4]+\"_16S_rRNA.fasta\", \"fasta\")\n",
    "                    #print(seq_16S)\n",
    "\n",
    "\n",
    "                #print\n",
    "            #print(record.features)\n",
    "    in_handle.close()\n",
    "\n",
    "                                        \n",
    "for root,dirs,files in os.walk(dir):\n",
    "    #print(dirs)\n",
    "    for tempFile in files:\n",
    "        if \"gff\" in tempFile:\n",
    "            path = os.path.join(root,tempFile)\n",
    "            FindrRNA(path, tempFile)\n",
    "            #print(path[:-4])\n",
    "            #print(dirs)\n",
    "            #print(root)\n",
    "            #print(tempFile[:-4])\n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "Created on Sat Nov  2 09:52:16 2019\n",
    "\n",
    "@author: Junyu\n",
    "\"\"\"\n",
    "\n",
    "import os\n",
    "from BCBio import GFF\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "\n",
    "dir = \"/home/junyuchen/Lab/Probe-Match/Root-WGS-Annotated/\"\n",
    "\n",
    "limit_info = dict(gff_type = [\"rRNA\"])\n",
    "        \n",
    "def FindrRNA(path, tempFile):\n",
    "    \n",
    "    in_handle = open(path)\n",
    "\n",
    "    for record in GFF.parse(in_handle, limit_info=limit_info):\n",
    "\n",
    "        if len(record.features) != 0:\n",
    "            for i in range(len(record.features)):\n",
    "\n",
    "                if '23S ribosomal RNA' in record.features[i].qualifiers[\"product\"][0]:\n",
    "                    #print(record.features[i])\n",
    "                    #print(record.id)\n",
    "                    #print(record.features[i].location.strand)\n",
    "                    #print(record.features[i].location)\n",
    "                    #print(record.features[i].location.start)\n",
    "                    #print(record.features[i].location.end)\n",
    "                    #print(type(record.features[i].location))\n",
    "                    #feature_seq = record.seq[record.features[i].location.start:record.features[i].location.end].reverse_complement()\n",
    "                    #print(feature_seq)\n",
    "                    '''\n",
    "                    seq_23S = SeqRecord(record.features[i].location.extract(record.seq),\\\n",
    "                                    id=record.features[i].id,\\\n",
    "                                    description=str(record.id +'-'+ record.features[i].qualifiers[\"product\"][0]))\n",
    "                    '''\n",
    "                    loc23 = int(record.features[i].location.end) - int(record.features[i].location.start)\n",
    "                    #print(loc)\n",
    "                    seq_23S = SeqRecord(record.features[i].location.extract(record.seq),\\\n",
    "                                    id=tempFile[:-4],\\\n",
    "                                    description=str( str(loc23)+ \"bp\"))\n",
    "                    #print(seq_23S)\n",
    "                    #print(record.features[i].location.start)\n",
    "                    #SeqIO.write(seq_23S, path[:-4]+\"_23S_rRNA.fasta\", \"fasta\")\n",
    "                    SeqIO.write(seq_23S, \"/home/junyuchen/Lab/Probe-Match/Root-rRNA3/\"+tempFile[:-4]+\"_23S_rRNA.fasta\", \"fasta\")\n",
    "                elif '16S ribosomal RNA' in record.features[i].qualifiers[\"product\"][0]:\n",
    "                    #print(record.features[i])\n",
    "                    #print(record.id)\n",
    "                    loc16 = int(record.features[i].location.end) - int(record.features[i].location.start)\n",
    "                    seq_16S = SeqRecord(record.features[i].location.extract(record.seq),\\\n",
    "                                    id=tempFile[:-4],\\\n",
    "                                    description=str(str(loc16)+\"bp\"))\n",
    "                \n",
    "                    #SeqIO.write(seq_16S, path[:-4]+\"_16S_rRNA.fasta\", \"fasta\")\n",
    "                    SeqIO.write(seq_16S, \"/home/junyuchen/Lab/Probe-Match/Root-rRNA3/\"+tempFile[:-4]+\"_16S_rRNA.fasta\", \"fasta\")\n",
    "                    #print(seq_16S)\n",
    "\n",
    "\n",
    "                #print\n",
    "            #print(record.features)\n",
    "    in_handle.close()\n",
    "\n",
    "                                        \n",
    "for root,dirs,files in os.walk(dir):\n",
    "    #print(dirs)\n",
    "    for tempFile in files:\n",
    "        if \"gff\" in tempFile:\n",
    "            path = os.path.join(root,tempFile)\n",
    "            FindrRNA(path, tempFile)\n",
    "            #print(path[:-4])\n",
    "            #print(dirs)\n",
    "            #print(root)\n",
    "            #print(tempFile[:-4])\n",
    "            \n",
    "            "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
